// @vendure/core/cli/job-queue/job-queue.service.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.JobQueueService = void 0;
const common_1 = require("@nestjs/common");
const config_1 = require("../config");
const constants_1 = require("./constants");
const job_buffer_service_1 = require("./job-buffer/job-buffer.service");
const job_queue_1 = require("./job-queue");
let JobQueueService = class JobQueueService {
    constructor(configService, jobBufferService) {
        this.configService = configService;
        this.jobBufferService = jobBufferService;
        this.queues = [];
        this.hasStarted = false;
    }
    get jobQueueStrategy() {
        return this.configService.jobQueueOptions.jobQueueStrategy;
    }
    onModuleDestroy() {
        this.hasStarted = false;
        return Promise.all(this.queues.map(q => q.stop()));
    }
    async createQueue(options) {
        if (this.configService.jobQueueOptions.prefix) {
            options = Object.assign(Object.assign({}, options), { name: `${this.configService.jobQueueOptions.prefix}${options.name}` });
        }
        const queue = new job_queue_1.JobQueue(options, this.jobQueueStrategy, this.jobBufferService);
        if (this.hasStarted && this.shouldStartQueue(queue.name)) {
            await queue.start();
        }
        this.queues.push(queue);
        return queue;
    }
    async start() {
        this.hasStarted = true;
        for (const queue of this.queues) {
            if (!queue.started && this.shouldStartQueue(queue.name)) {
                config_1.Logger.info(`Starting queue: ${queue.name}`, constants_1.loggerCtx);
                await queue.start();
            }
        }
    }
    addBuffer(buffer) {
        this.jobBufferService.addBuffer(buffer);
    }
    removeBuffer(buffer) {
        this.jobBufferService.removeBuffer(buffer);
    }
    bufferSize(...forBuffers) {
        return this.jobBufferService.bufferSize(forBuffers);
    }
    flush(...forBuffers) {
        return this.jobBufferService.flush(forBuffers);
    }
    getJobQueues() {
        return this.queues.map(queue => ({
            name: queue.name,
            running: queue.started,
        }));
    }
    shouldStartQueue(queueName) {
        if (this.configService.jobQueueOptions.activeQueues.length > 0) {
            if (!this.configService.jobQueueOptions.activeQueues.includes(queueName)) {
                return false;
            }
        }
        return true;
    }
};
JobQueueService = __decorate([
    common_1.Injectable(),
    __metadata("design:paramtypes", [config_1.ConfigService, job_buffer_service_1.JobBufferService])
], JobQueueService);
exports.JobQueueService = JobQueueService;
