// @vendure/admin-ui/esm2015/core/shared/components/affixed-input/affixed-input.component.js
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
/**
 * A wrapper around an <input> element which adds a prefix and/or a suffix element.
 */
export class AffixedInputComponent {
}
AffixedInputComponent.decorators = [
    { type: Component, args: [{
                selector: 'vdr-affixed-input',
                template: "<div [class.has-prefix]=\"!!prefix\" [class.has-suffix]=\"!!suffix\">\r\n    <ng-content></ng-content>\r\n</div>\r\n<div class=\"affix prefix\" *ngIf=\"prefix\">{{ prefix }}</div>\r\n<div class=\"affix suffix\" *ngIf=\"suffix\">{{ suffix }}</div>\r\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [":host{display:inline-flex}.affix{color:var(--color-grey-800);display:flex;align-items:center;background-color:var(--color-grey-200);border:1px solid var(--color-component-border-300);top:1px;padding:3px;line-height:.58333rem;transition:border .2s}::ng-deep .has-prefix input{border-top-left-radius:0!important;border-bottom-left-radius:0!important}.prefix{order:-1;border-radius:3px 0 0 3px;border-right:none}::ng-deep .has-suffix input{border-top-right-radius:0!important;border-bottom-right-radius:0!important}.suffix{border-radius:0 3px 3px 0;border-left:none}\n"]
            },] }
];
AffixedInputComponent.propDecorators = {
    prefix: [{ type: Input }],
    suffix: [{ type: Input }]
};
