// @vendure/admin-ui/esm2015/customer/customer.routes.js
import { marker as _ } from '@biesbjerg/ngx-translate-extract-marker';
import { CanDeactivateDetailGuard, createResolveData, detailBreadcrumb, } from '@vendure/admin-ui/core';
import { CustomerDetailComponent } from './components/customer-detail/customer-detail.component';
import { CustomerGroupListComponent } from './components/customer-group-list/customer-group-list.component';
import { CustomerListComponent } from './components/customer-list/customer-list.component';
import { CustomerResolver } from './providers/routing/customer-resolver';
const ɵ0 = {
    breadcrumb: _('breadcrumb.customers'),
}, ɵ1 = {
    breadcrumb: customerBreadcrumb,
}, ɵ2 = {
    breadcrumb: _('breadcrumb.customer-groups'),
};
export const customerRoutes = [
    {
        path: 'customers',
        component: CustomerListComponent,
        pathMatch: '',
        data: ɵ0,
    },
    {
        path: 'customers/:id',
        component: CustomerDetailComponent,
        resolve: createResolveData(CustomerResolver),
        canDeactivate: [CanDeactivateDetailGuard],
        data: ɵ1,
    },
    {
        path: 'groups',
        component: CustomerGroupListComponent,
        data: ɵ2,
    },
];
export function customerBreadcrumb(data, params) {
    return detailBreadcrumb({
        entity: data.entity,
        id: params.id,
        breadcrumbKey: 'breadcrumb.customers',
        getName: (customer) => `${customer.firstName} ${customer.lastName}`,
        route: 'customers',
    });
}
export { ɵ0, ɵ1, ɵ2 };
