// @vendure/admin-ui/esm2015/core/shared/components/items-per-page-controls/items-per-page-controls.component.js
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output } from '@angular/core';
/**
 * A control for setting the number of items per page in a paginated list.
 */
export class ItemsPerPageControlsComponent {
    constructor() {
        this.itemsPerPage = 10;
        this.itemsPerPageChange = new EventEmitter();
    }
}
ItemsPerPageControlsComponent.decorators = [
    { type: Component, args: [{
                selector: 'vdr-items-per-page-controls',
                template: "<div class=\"select\">\r\n    <select [ngModel]=\"itemsPerPage\" (change)=\"itemsPerPageChange.emit($event.target.value)\">\r\n        <option [value]=\"10\">{{ 'common.items-per-page-option' | translate: { count: 10 } }}</option>\r\n        <option [value]=\"25\">{{ 'common.items-per-page-option' | translate: { count: 25 } }}</option>\r\n        <option [value]=\"50\">{{ 'common.items-per-page-option' | translate: { count: 50 } }}</option>\r\n        <option [value]=\"100\">{{ 'common.items-per-page-option' | translate: { count: 100 } }}</option>\r\n    </select>\r\n</div>\r\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [""]
            },] }
];
ItemsPerPageControlsComponent.propDecorators = {
    itemsPerPage: [{ type: Input }],
    itemsPerPageChange: [{ type: Output }]
};
