// @vendure/admin-ui/esm2015/core/shared/components/action-bar/action-bar.component.js
import { Component, ContentChild, Input } from '@angular/core';
export class ActionBarLeftComponent {
    constructor() {
        this.grow = false;
    }
}
ActionBarLeftComponent.decorators = [
    { type: Component, args: [{
                selector: 'vdr-ab-left',
                template: `
        <ng-content></ng-content>
    `
            },] }
];
ActionBarLeftComponent.propDecorators = {
    grow: [{ type: Input }]
};
export class ActionBarRightComponent {
    constructor() {
        this.grow = false;
    }
}
ActionBarRightComponent.decorators = [
    { type: Component, args: [{
                selector: 'vdr-ab-right',
                template: `
        <ng-content></ng-content>
    `
            },] }
];
ActionBarRightComponent.propDecorators = {
    grow: [{ type: Input }]
};
export class ActionBarComponent {
}
ActionBarComponent.decorators = [
    { type: Component, args: [{
                selector: 'vdr-action-bar',
                template: "<div class=\"left-content\" [class.grow]=\"left?.grow\"><ng-content select=\"vdr-ab-left\"></ng-content></div>\r\n<div class=\"right-content\" [class.grow]=\"right?.grow\"><ng-content select=\"vdr-ab-right\"></ng-content></div>\r\n",
                styles: [":host{display:flex;justify-content:space-between;align-items:baseline;background-color:var(--color-component-bg-100);position:sticky;top:-24px;z-index:25;border-bottom:1px solid var(--color-component-border-200)}:host>.grow{flex:1}\n"]
            },] }
];
ActionBarComponent.propDecorators = {
    left: [{ type: ContentChild, args: [ActionBarLeftComponent,] }],
    right: [{ type: ContentChild, args: [ActionBarRightComponent,] }]
};
