// @vendure/admin-ui/esm2015/core/providers/health-check/health-check.service.js
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { merge, of, Subject, timer } from 'rxjs';
import { catchError, map, shareReplay, switchMap, throttleTime } from 'rxjs/operators';
import { getServerLocation } from '../../data/utils/get-server-location';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
export class HealthCheckService {
    constructor(httpClient) {
        this.httpClient = httpClient;
        this.pollingDelayMs = 60 * 1000;
        this._refresh = new Subject();
        this.healthCheckEndpoint = getServerLocation() + '/health';
        const refresh$ = this._refresh.pipe(throttleTime(1000));
        const result$ = merge(timer(0, this.pollingDelayMs), refresh$).pipe(switchMap(() => this.checkHealth()), shareReplay(1));
        this.status$ = result$.pipe(map(res => res.status));
        this.details$ = result$.pipe(map(res => Object.keys(res.details).map(key => {
            return { key, result: res.details[key] };
        })));
        this.lastCheck$ = result$.pipe(map(res => res.lastChecked));
    }
    refresh() {
        this._refresh.next();
    }
    checkHealth() {
        return this.httpClient.get(this.healthCheckEndpoint).pipe(catchError(err => of(err.error)), map(res => (Object.assign(Object.assign({}, res), { lastChecked: new Date() }))));
    }
}
HealthCheckService.ɵprov = i0.ɵɵdefineInjectable({ factory: function HealthCheckService_Factory() { return new HealthCheckService(i0.ɵɵinject(i1.HttpClient)); }, token: HealthCheckService, providedIn: "root" });
HealthCheckService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
HealthCheckService.ctorParameters = () => [
    { type: HttpClient }
];
