// @vendure/admin-ui/esm2015/order/components/line-refunds/line-refunds.component.js
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
export class LineRefundsComponent {
    getRefundedCount() {
        var _a, _b;
        const refunds = (_b = (_a = this.payments) === null || _a === void 0 ? void 0 : _a.reduce((all, payment) => [...all, ...payment.refunds], [])) !== null && _b !== void 0 ? _b : [];
        return this.line.items.filter(i => {
            if (i.refundId === null && !i.cancelled) {
                return false;
            }
            if (i.refundId) {
                const refund = refunds.find(r => r.id === i.refundId);
                if ((refund === null || refund === void 0 ? void 0 : refund.state) === 'Failed') {
                    return false;
                }
                else {
                    return true;
                }
            }
            return false;
        }).length;
    }
}
LineRefundsComponent.decorators = [
    { type: Component, args: [{
                selector: 'vdr-line-refunds',
                template: "<span *ngIf=\"getRefundedCount()\" [title]=\"'order.refunded-count' | translate: { count: getRefundedCount() }\">\r\n    <clr-icon shape=\"redo\" class=\"is-solid\" dir=\"down\"></clr-icon>\r\n</span>\r\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [":host{color:var(--color-error-500)}\n"]
            },] }
];
LineRefundsComponent.propDecorators = {
    line: [{ type: Input }],
    payments: [{ type: Input }]
};
