// @vendure/admin-ui/esm2015/core/shared/dynamic-form-inputs/text-form-input/text-form-input.component.js
import { ChangeDetectionStrategy, Component } from '@angular/core';
/**
 * @description
 * Uses a regular text form input. This is the default input for `string` and `localeString` type fields.
 *
 * @docsCategory custom-input-components
 * @docsPage default-inputs
 */
export class TextFormInputComponent {
    get prefix() {
        var _a;
        return ((_a = this.config.ui) === null || _a === void 0 ? void 0 : _a.prefix) || this.config.prefix;
    }
    get suffix() {
        var _a;
        return ((_a = this.config.ui) === null || _a === void 0 ? void 0 : _a.suffix) || this.config.suffix;
    }
}
TextFormInputComponent.id = 'text-form-input';
TextFormInputComponent.decorators = [
    { type: Component, args: [{
                selector: 'vdr-text-form-input',
                template: "<vdr-affixed-input\r\n    [suffix]=\"suffix\"\r\n    [prefix]=\"prefix\"\r\n>\r\n    <input type=\"text\" [readonly]=\"readonly\" [formControl]=\"formControl\" />\r\n</vdr-affixed-input>\r\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: ["input{width:100%}\n"]
            },] }
];
