// @vendure/core/cli/common/ttl-cache.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TtlCache = void 0;
class TtlCache {
    constructor(config) {
        this.cache = new Map();
        this.ttl = 30 * 1000;
        this.cacheSize = 1000;
        if (config === null || config === void 0 ? void 0 : config.ttl) {
            this.ttl = config.ttl;
        }
        if (config === null || config === void 0 ? void 0 : config.cacheSize) {
            this.cacheSize = config.cacheSize;
        }
    }
    get(key) {
        const hit = this.cache.get(key);
        const now = new Date().getTime();
        if (hit) {
            if (now < hit.expires) {
                return hit.value;
            }
            else {
                this.cache.delete(key);
            }
        }
    }
    set(key, value) {
        if (this.cache.has(key)) {
            this.cache.delete(key);
        }
        else if (this.cache.size === this.cacheSize) {
            this.cache.delete(this.first());
        }
        this.cache.set(key, {
            value,
            expires: new Date().getTime() + this.ttl,
        });
    }
    delete(key) {
        this.cache.delete(key);
    }
    first() {
        return this.cache.keys().next().value;
    }
}
exports.TtlCache = TtlCache;
