// @vendure/core/cli/config/payment/dummy-payment-method-handler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.dummyPaymentHandler = void 0;
const generated_types_1 = require("@vendure/common/lib/generated-types");
const payment_method_handler_1 = require("./payment-method-handler");
exports.dummyPaymentHandler = new payment_method_handler_1.PaymentMethodHandler({
    code: 'dummy-payment-handler',
    description: [
        {
            languageCode: generated_types_1.LanguageCode.en,
            value: 'A dummy payment provider intended for testing and development only.',
        },
    ],
    args: {
        automaticSettle: {
            type: 'boolean',
            label: [
                {
                    languageCode: generated_types_1.LanguageCode.en,
                    value: 'Authorize and settle in 1 step',
                },
            ],
            description: [
                {
                    languageCode: generated_types_1.LanguageCode.en,
                    value: 'If enabled, Payments will be created in the "Settled" state.',
                },
            ],
            required: true,
            defaultValue: false,
        },
    },
    createPayment: async (ctx, order, amount, args, metadata) => {
        if (metadata.shouldDecline) {
            return {
                amount,
                state: 'Declined',
                metadata: {
                    errorMessage: 'Simulated decline',
                },
            };
        }
        else if (metadata.shouldError) {
            return {
                amount,
                state: 'Error',
                errorMessage: 'Simulated error',
                metadata: {
                    errorMessage: 'Simulated error',
                },
            };
        }
        else {
            return {
                amount,
                state: args.automaticSettle ? 'Settled' : 'Authorized',
                transactionId: Math.random().toString(36).substr(3),
                metadata,
            };
        }
    },
    settlePayment: async (ctx, order, payment, args) => {
        if (payment.metadata.shouldErrorOnSettle) {
            return {
                success: false,
                errorMessage: 'Simulated settlement error',
            };
        }
        return {
            success: true,
        };
    },
});
