// @vendure/admin-ui/esm2015/settings/providers/routing/channel-resolver.js
import { Injectable } from '@angular/core';
import { Router } from '@angular/router';
import { BaseEntityResolver } from '@vendure/admin-ui/core';
import { CurrencyCode } from '@vendure/admin-ui/core';
import { getDefaultUiLanguage } from '@vendure/admin-ui/core';
import { DataService } from '@vendure/admin-ui/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
import * as i2 from "@vendure/admin-ui/core";
/**
 * Resolves the id from the path into a Customer entity.
 */
export class ChannelResolver extends BaseEntityResolver {
    constructor(router, dataService) {
        super(router, {
            __typename: 'Channel',
            id: '',
            createdAt: '',
            updatedAt: '',
            code: '',
            token: '',
            pricesIncludeTax: false,
            currencyCode: CurrencyCode.USD,
            defaultLanguageCode: getDefaultUiLanguage(),
            defaultShippingZone: {},
            defaultTaxZone: {},
        }, (id) => dataService.settings.getChannel(id).mapStream((data) => data.channel));
    }
}
ChannelResolver.ɵprov = i0.ɵɵdefineInjectable({ factory: function ChannelResolver_Factory() { return new ChannelResolver(i0.ɵɵinject(i1.Router), i0.ɵɵinject(i2.DataService)); }, token: ChannelResolver, providedIn: "root" });
ChannelResolver.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
ChannelResolver.ctorParameters = () => [
    { type: Router },
    { type: DataService }
];
