// @vendure/admin-ui/esm2015/settings/providers/routing/country-resolver.js
import { Injectable } from '@angular/core';
import { Router } from '@angular/router';
import { BaseEntityResolver } from '@vendure/admin-ui/core';
import { DataService } from '@vendure/admin-ui/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
import * as i2 from "@vendure/admin-ui/core";
/**
 * Resolves the id from the path into a Customer entity.
 */
export class CountryResolver extends BaseEntityResolver {
    constructor(router, dataService) {
        super(router, {
            __typename: 'Country',
            id: '',
            createdAt: '',
            updatedAt: '',
            code: '',
            name: '',
            enabled: false,
            translations: [],
        }, id => dataService.settings.getCountry(id).mapStream(data => data.country));
    }
}
CountryResolver.ɵprov = i0.ɵɵdefineInjectable({ factory: function CountryResolver_Factory() { return new CountryResolver(i0.ɵɵinject(i1.Router), i0.ɵɵinject(i2.DataService)); }, token: CountryResolver, providedIn: "root" });
CountryResolver.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
CountryResolver.ctorParameters = () => [
    { type: Router },
    { type: DataService }
];
