// @vendure/admin-ui/esm2015/settings/providers/routing/payment-method-resolver.js
import { Injectable } from '@angular/core';
import { Router } from '@angular/router';
import { BaseEntityResolver } from '@vendure/admin-ui/core';
import { DataService } from '@vendure/admin-ui/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
import * as i2 from "@vendure/admin-ui/core";
/**
 * Resolves the id from the path into a Customer entity.
 */
export class PaymentMethodResolver extends BaseEntityResolver {
    constructor(router, dataService) {
        super(router, {
            __typename: 'PaymentMethod',
            id: '',
            createdAt: '',
            updatedAt: '',
            name: '',
            code: '',
            description: '',
            enabled: true,
            checker: undefined,
            handler: undefined,
        }, id => dataService.settings.getPaymentMethod(id).mapStream(data => data.paymentMethod));
    }
}
PaymentMethodResolver.ɵprov = i0.ɵɵdefineInjectable({ factory: function PaymentMethodResolver_Factory() { return new PaymentMethodResolver(i0.ɵɵinject(i1.Router), i0.ɵɵinject(i2.DataService)); }, token: PaymentMethodResolver, providedIn: "root" });
PaymentMethodResolver.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
PaymentMethodResolver.ctorParameters = () => [
    { type: Router },
    { type: DataService }
];
