// @vendure/admin-ui/esm2015/settings/providers/routing/role-resolver.js
import { Injectable } from '@angular/core';
import { Router } from '@angular/router';
import { BaseEntityResolver } from '@vendure/admin-ui/core';
import { DataService } from '@vendure/admin-ui/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
import * as i2 from "@vendure/admin-ui/core";
export class RoleResolver extends BaseEntityResolver {
    constructor(router, dataService) {
        super(router, {
            __typename: 'Role',
            id: '',
            createdAt: '',
            updatedAt: '',
            code: '',
            description: '',
            permissions: [],
            channels: [],
        }, id => dataService.administrator.getRole(id).mapStream(data => data.role));
    }
}
RoleResolver.ɵprov = i0.ɵɵdefineInjectable({ factory: function RoleResolver_Factory() { return new RoleResolver(i0.ɵɵinject(i1.Router), i0.ɵɵinject(i2.DataService)); }, token: RoleResolver, providedIn: "root" });
RoleResolver.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
RoleResolver.ctorParameters = () => [
    { type: Router },
    { type: DataService }
];
