// @vendure/admin-ui/esm2015/settings/providers/routing/shipping-method-resolver.js
import { Injectable } from '@angular/core';
import { Router } from '@angular/router';
import { BaseEntityResolver } from '@vendure/admin-ui/core';
import { DataService } from '@vendure/admin-ui/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
import * as i2 from "@vendure/admin-ui/core";
/**
 * Resolves the id from the path into a Customer entity.
 */
export class ShippingMethodResolver extends BaseEntityResolver {
    constructor(router, dataService) {
        super(router, {
            __typename: 'ShippingMethod',
            createdAt: '',
            updatedAt: '',
            id: '',
            code: '',
            name: '',
            description: '',
            fulfillmentHandlerCode: undefined,
            checker: undefined,
            calculator: undefined,
            translations: [],
        }, id => dataService.shippingMethod.getShippingMethod(id).mapStream(data => data.shippingMethod));
    }
}
ShippingMethodResolver.ɵprov = i0.ɵɵdefineInjectable({ factory: function ShippingMethodResolver_Factory() { return new ShippingMethodResolver(i0.ɵɵinject(i1.Router), i0.ɵɵinject(i2.DataService)); }, token: ShippingMethodResolver, providedIn: "root" });
ShippingMethodResolver.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
ShippingMethodResolver.ctorParameters = () => [
    { type: Router },
    { type: DataService }
];
