// @vendure/admin-ui/esm2015/settings/providers/routing/tax-category-resolver.js
import { Injectable } from '@angular/core';
import { Router } from '@angular/router';
import { BaseEntityResolver, DataService } from '@vendure/admin-ui/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
import * as i2 from "@vendure/admin-ui/core";
/**
 * Resolves the id from the path into a Customer entity.
 */
export class TaxCategoryResolver extends BaseEntityResolver {
    constructor(router, dataService) {
        super(router, {
            __typename: 'TaxCategory',
            id: '',
            createdAt: '',
            updatedAt: '',
            name: '',
            isDefault: false,
        }, id => dataService.settings.getTaxCategory(id).mapStream(data => data.taxCategory));
    }
}
TaxCategoryResolver.ɵprov = i0.ɵɵdefineInjectable({ factory: function TaxCategoryResolver_Factory() { return new TaxCategoryResolver(i0.ɵɵinject(i1.Router), i0.ɵɵinject(i2.DataService)); }, token: TaxCategoryResolver, providedIn: "root" });
TaxCategoryResolver.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
TaxCategoryResolver.ctorParameters = () => [
    { type: Router },
    { type: DataService }
];
