// @vendure/admin-ui/esm2015/settings/providers/routing/tax-rate-resolver.js
import { Injectable } from '@angular/core';
import { Router } from '@angular/router';
import { BaseEntityResolver, DataService } from '@vendure/admin-ui/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
import * as i2 from "@vendure/admin-ui/core";
/**
 * Resolves the id from the path into a Customer entity.
 */
export class TaxRateResolver extends BaseEntityResolver {
    constructor(router, dataService) {
        super(router, {
            __typename: 'TaxRate',
            id: '',
            createdAt: '',
            updatedAt: '',
            name: '',
            value: 0,
            enabled: true,
            category: {},
            zone: {},
            customerGroup: null,
        }, id => dataService.settings.getTaxRate(id).mapStream(data => data.taxRate));
    }
}
TaxRateResolver.ɵprov = i0.ɵɵdefineInjectable({ factory: function TaxRateResolver_Factory() { return new TaxRateResolver(i0.ɵɵinject(i1.Router), i0.ɵɵinject(i2.DataService)); }, token: TaxRateResolver, providedIn: "root" });
TaxRateResolver.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
TaxRateResolver.ctorParameters = () => [
    { type: Router },
    { type: DataService }
];
