// @vendure/admin-ui/esm2015/core/shared/components/rich-text-editor/link-dialog/link-dialog.component.js
import { ChangeDetectionStrategy, Component } from '@angular/core';
import { FormControl, FormGroup, Validators } from '@angular/forms';
export class LinkDialogComponent {
    ngOnInit() {
        this.form = new FormGroup({
            href: new FormControl(this.existing ? this.existing.href : '', Validators.required),
            title: new FormControl(this.existing ? this.existing.title : ''),
        });
    }
    remove() {
        this.resolveWith({
            title: '',
            href: '',
        });
    }
    select() {
        this.resolveWith(this.form.value);
    }
}
LinkDialogComponent.decorators = [
    { type: Component, args: [{
                selector: 'vdr-link-dialog',
                template: "<form [formGroup]=\"form\">\r\n    <vdr-form-field [label]=\"'editor.link-href' | translate\" for=\"href\">\r\n        <input id=\"href\" type=\"text\" formControlName=\"href\" />\r\n    </vdr-form-field>\r\n    <vdr-form-field [label]=\"'editor.link-title' | translate\" for=\"title\">\r\n        <input id=\"title\" type=\"text\" formControlName=\"title\" />\r\n    </vdr-form-field>\r\n</form>\r\n<ng-template vdrDialogButtons>\r\n    <button type=\"button\" class=\"btn btn-secondary\" (click)=\"remove()\" *ngIf=\"existing\">\r\n        <clr-icon shape=\"unlink\"></clr-icon> {{ 'editor.remove-link' | translate }}\r\n    </button>\r\n    <button type=\"submit\" (click)=\"select()\" class=\"btn btn-primary\" [disabled]=\"form.invalid\">\r\n        {{ 'editor.set-link' | translate }}\r\n    </button>\r\n</ng-template>\r\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [""]
            },] }
];
