// @vendure/admin-ui/esm2015/core/providers/overlay-host/overlay-host.service.js
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * The OverlayHostService is used to get a reference to the ViewConainerRef of the
 * OverlayHost component, so that other components may insert components & elements
 * into the DOM at that point.
 */
export class OverlayHostService {
    constructor() {
        this.promiseResolveFns = [];
    }
    /**
     * Used to pass in the ViewContainerRed from the OverlayHost component.
     * Should not be used by any other component.
     */
    registerHostView(viewContainerRef) {
        this.hostView = viewContainerRef;
        if (0 < this.promiseResolveFns.length) {
            this.resolveHostView();
        }
    }
    /**
     * Returns a promise which resolves to the ViewContainerRef of the OverlayHost
     * component. This can then be used to insert components and elements into the
     * DOM at that point.
     */
    getHostView() {
        return new Promise((resolve) => {
            this.promiseResolveFns.push(resolve);
            if (this.hostView !== undefined) {
                this.resolveHostView();
            }
        });
    }
    resolveHostView() {
        this.promiseResolveFns.forEach(resolve => resolve(this.hostView));
        this.promiseResolveFns = [];
    }
}
OverlayHostService.ɵprov = i0.ɵɵdefineInjectable({ factory: function OverlayHostService_Factory() { return new OverlayHostService(); }, token: OverlayHostService, providedIn: "root" });
OverlayHostService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
