// @vendure/admin-ui/esm2015/core/shared/dynamic-form-inputs/date-form-input/date-form-input.component.js
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
/**
 * @description
 * Allows selection of a datetime. Default input for `datetime` type fields.
 *
 * @docsCategory custom-input-components
 * @docsPage default-inputs
 */
export class DateFormInputComponent {
    get min() {
        var _a;
        return ((_a = this.config.ui) === null || _a === void 0 ? void 0 : _a.min) || this.config.min;
    }
    get max() {
        var _a;
        return ((_a = this.config.ui) === null || _a === void 0 ? void 0 : _a.max) || this.config.max;
    }
    get yearRange() {
        var _a;
        return ((_a = this.config.ui) === null || _a === void 0 ? void 0 : _a.yearRange) || this.config.yearRange;
    }
}
DateFormInputComponent.id = 'date-form-input';
DateFormInputComponent.decorators = [
    { type: Component, args: [{
                selector: 'vdr-date-form-input',
                template: "<vdr-datetime-picker\r\n    [formControl]=\"formControl\"\r\n    [min]=\"min\"\r\n    [max]=\"max\"\r\n    [yearRange]=\"yearRange\"\r\n    [readonly]=\"readonly\"\r\n>\r\n</vdr-datetime-picker>\r\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [""]
            },] }
];
DateFormInputComponent.propDecorators = {
    readonly: [{ type: Input }]
};
