// @vendure/admin-ui/esm2015/catalog/catalog.routes.js
import { marker as _ } from '@biesbjerg/ngx-translate-extract-marker';
import { CanDeactivateDetailGuard, createResolveData, detailBreadcrumb, } from '@vendure/admin-ui/core';
import { map } from 'rxjs/operators';
import { AssetDetailComponent } from './components/asset-detail/asset-detail.component';
import { AssetListComponent } from './components/asset-list/asset-list.component';
import { CollectionDetailComponent } from './components/collection-detail/collection-detail.component';
import { CollectionListComponent } from './components/collection-list/collection-list.component';
import { FacetDetailComponent } from './components/facet-detail/facet-detail.component';
import { FacetListComponent } from './components/facet-list/facet-list.component';
import { ProductDetailComponent } from './components/product-detail/product-detail.component';
import { ProductListComponent } from './components/product-list/product-list.component';
import { ProductOptionsEditorComponent } from './components/product-options-editor/product-options-editor.component';
import { ProductVariantsEditorComponent } from './components/product-variants-editor/product-variants-editor.component';
import { AssetResolver } from './providers/routing/asset-resolver';
import { CollectionResolver } from './providers/routing/collection-resolver';
import { FacetResolver } from './providers/routing/facet-resolver';
import { ProductResolver } from './providers/routing/product-resolver';
import { ProductVariantsResolver } from './providers/routing/product-variants-resolver';
const ɵ0 = {
    breadcrumb: _('breadcrumb.products'),
}, ɵ1 = {
    breadcrumb: productBreadcrumb,
}, ɵ2 = {
    breadcrumb: productVariantEditorBreadcrumb,
}, ɵ3 = {
    breadcrumb: productOptionsEditorBreadcrumb,
}, ɵ4 = {
    breadcrumb: _('breadcrumb.facets'),
}, ɵ5 = {
    breadcrumb: facetBreadcrumb,
}, ɵ6 = {
    breadcrumb: _('breadcrumb.collections'),
}, ɵ7 = {
    breadcrumb: collectionBreadcrumb,
}, ɵ8 = {
    breadcrumb: _('breadcrumb.assets'),
}, ɵ9 = {
    breadcrumb: assetBreadcrumb,
};
export const catalogRoutes = [
    {
        path: 'products',
        component: ProductListComponent,
        data: ɵ0,
    },
    {
        path: 'products/:id',
        component: ProductDetailComponent,
        resolve: createResolveData(ProductResolver),
        canDeactivate: [CanDeactivateDetailGuard],
        data: ɵ1,
    },
    {
        path: 'products/:id/manage-variants',
        component: ProductVariantsEditorComponent,
        resolve: createResolveData(ProductVariantsResolver),
        canDeactivate: [CanDeactivateDetailGuard],
        data: ɵ2,
    },
    {
        path: 'products/:id/options',
        component: ProductOptionsEditorComponent,
        resolve: createResolveData(ProductVariantsResolver),
        canDeactivate: [CanDeactivateDetailGuard],
        data: ɵ3,
    },
    {
        path: 'facets',
        component: FacetListComponent,
        data: ɵ4,
    },
    {
        path: 'facets/:id',
        component: FacetDetailComponent,
        resolve: createResolveData(FacetResolver),
        canDeactivate: [CanDeactivateDetailGuard],
        data: ɵ5,
    },
    {
        path: 'collections',
        component: CollectionListComponent,
        data: ɵ6,
    },
    {
        path: 'collections/:id',
        component: CollectionDetailComponent,
        resolve: createResolveData(CollectionResolver),
        canDeactivate: [CanDeactivateDetailGuard],
        data: ɵ7,
    },
    {
        path: 'assets',
        component: AssetListComponent,
        data: ɵ8,
    },
    {
        path: 'assets/:id',
        component: AssetDetailComponent,
        resolve: createResolveData(AssetResolver),
        data: ɵ9,
    },
];
export function productBreadcrumb(data, params) {
    return detailBreadcrumb({
        entity: data.entity,
        id: params.id,
        breadcrumbKey: 'breadcrumb.products',
        getName: product => product.name,
        route: 'products',
    });
}
export function productVariantEditorBreadcrumb(data, params) {
    return data.entity.pipe(map((entity) => {
        return [
            {
                label: _('breadcrumb.products'),
                link: ['../', 'products'],
            },
            {
                label: `${entity.name}`,
                link: ['../', 'products', params.id, { tab: 'variants' }],
            },
            {
                label: _('breadcrumb.manage-variants'),
                link: ['manage-variants'],
            },
        ];
    }));
}
export function productOptionsEditorBreadcrumb(data, params) {
    return data.entity.pipe(map((entity) => {
        return [
            {
                label: _('breadcrumb.products'),
                link: ['../', 'products'],
            },
            {
                label: `${entity.name}`,
                link: ['../', 'products', params.id, { tab: 'variants' }],
            },
            {
                label: _('breadcrumb.product-options'),
                link: ['options'],
            },
        ];
    }));
}
export function facetBreadcrumb(data, params) {
    return detailBreadcrumb({
        entity: data.entity,
        id: params.id,
        breadcrumbKey: 'breadcrumb.facets',
        getName: facet => facet.name,
        route: 'facets',
    });
}
export function collectionBreadcrumb(data, params) {
    return detailBreadcrumb({
        entity: data.entity,
        id: params.id,
        breadcrumbKey: 'breadcrumb.collections',
        getName: collection => collection.name,
        route: 'collections',
    });
}
export function assetBreadcrumb(data, params) {
    return detailBreadcrumb({
        entity: data.entity,
        id: params.id,
        breadcrumbKey: 'breadcrumb.assets',
        getName: asset => asset.name,
        route: 'assets',
    });
}
export { ɵ0, ɵ1, ɵ2, ɵ3, ɵ4, ɵ5, ɵ6, ɵ7, ɵ8, ɵ9 };
