// @vendure/core/cli/config/promotion/promotion-action.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PromotionShippingAction = exports.PromotionOrderAction = exports.PromotionItemAction = exports.PromotionAction = void 0;
const pick_1 = require("@vendure/common/lib/pick");
const configurable_operation_1 = require("../../common/configurable-operation");
class PromotionAction extends configurable_operation_1.ConfigurableOperationDef {
    constructor(config) {
        super(config);
        this.priorityValue = config.priorityValue || 0;
        this.conditions = config.conditions;
    }
}
exports.PromotionAction = PromotionAction;
class PromotionItemAction extends PromotionAction {
    constructor(config) {
        super(config);
        this.executeFn = config.execute;
    }
    execute(ctx, orderItem, orderLine, args, state) {
        const actionState = this.conditions
            ? pick_1.pick(state, this.conditions.map(c => c.code))
            : {};
        return this.executeFn(ctx, orderItem, orderLine, this.argsArrayToHash(args), actionState);
    }
}
exports.PromotionItemAction = PromotionItemAction;
class PromotionOrderAction extends PromotionAction {
    constructor(config) {
        super(config);
        this.executeFn = config.execute;
    }
    execute(ctx, order, args, state) {
        const actionState = this.conditions ? pick_1.pick(state, this.conditions.map(c => c.code)) : {};
        return this.executeFn(ctx, order, this.argsArrayToHash(args), actionState);
    }
}
exports.PromotionOrderAction = PromotionOrderAction;
class PromotionShippingAction extends PromotionAction {
    constructor(config) {
        super(config);
        this.executeFn = config.execute;
    }
    execute(ctx, shippingLine, order, args, state) {
        const actionState = this.conditions ? pick_1.pick(state, this.conditions.map(c => c.code)) : {};
        return this.executeFn(ctx, shippingLine, order, this.argsArrayToHash(args), actionState);
    }
}
exports.PromotionShippingAction = PromotionShippingAction;
