// @vendure/admin-ui/esm2015/core/shared/dynamic-form-inputs/number-form-input/number-form-input.component.js
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
/**
 * @description
 * Displays a number input. Default input for `int` and `float` type fields.
 *
 * @docsCategory custom-input-components
 * @docsPage default-inputs
 */
export class NumberFormInputComponent {
    get prefix() {
        var _a;
        return ((_a = this.config.ui) === null || _a === void 0 ? void 0 : _a.prefix) || this.config.prefix;
    }
    get suffix() {
        var _a;
        return ((_a = this.config.ui) === null || _a === void 0 ? void 0 : _a.suffix) || this.config.suffix;
    }
    get min() {
        var _a;
        return ((_a = this.config.ui) === null || _a === void 0 ? void 0 : _a.min) || this.config.min;
    }
    get max() {
        var _a;
        return ((_a = this.config.ui) === null || _a === void 0 ? void 0 : _a.max) || this.config.max;
    }
    get step() {
        var _a;
        return ((_a = this.config.ui) === null || _a === void 0 ? void 0 : _a.step) || this.config.step;
    }
}
NumberFormInputComponent.id = 'number-form-input';
NumberFormInputComponent.decorators = [
    { type: Component, args: [{
                selector: 'vdr-number-form-input',
                template: "<vdr-affixed-input\r\n    [suffix]=\"suffix\"\r\n    [prefix]=\"prefix\"\r\n>\r\n    <input\r\n        type=\"number\"\r\n        [readonly]=\"readonly\"\r\n        [min]=\"min\"\r\n        [max]=\"max\"\r\n        [step]=\"step\"\r\n        [formControl]=\"formControl\"\r\n    />\r\n</vdr-affixed-input>\r\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [""]
            },] }
];
NumberFormInputComponent.propDecorators = {
    readonly: [{ type: Input }]
};
