// @vendure/admin-ui/esm2015/marketing/marketing.routes.js
import { marker as _ } from '@biesbjerg/ngx-translate-extract-marker';
import { CanDeactivateDetailGuard, createResolveData, detailBreadcrumb } from '@vendure/admin-ui/core';
import { PromotionDetailComponent } from './components/promotion-detail/promotion-detail.component';
import { PromotionListComponent } from './components/promotion-list/promotion-list.component';
import { PromotionResolver } from './providers/routing/promotion-resolver';
const ɵ0 = {
    breadcrumb: _('breadcrumb.promotions'),
}, ɵ1 = {
    breadcrumb: promotionBreadcrumb,
};
export const marketingRoutes = [
    {
        path: 'promotions',
        component: PromotionListComponent,
        data: ɵ0,
    },
    {
        path: 'promotions/:id',
        component: PromotionDetailComponent,
        resolve: createResolveData(PromotionResolver),
        canDeactivate: [CanDeactivateDetailGuard],
        data: ɵ1,
    },
];
export function promotionBreadcrumb(data, params) {
    return detailBreadcrumb({
        entity: data.entity,
        id: params.id,
        breadcrumbKey: 'breadcrumb.promotions',
        getName: promotion => promotion.name,
        route: 'promotions',
    });
}
export { ɵ0, ɵ1 };
