// @vendure/admin-ui/esm2015/catalog/providers/routing/asset-resolver.js
import { Injectable } from '@angular/core';
import { Router } from '@angular/router';
import { AssetType, BaseEntityResolver } from '@vendure/admin-ui/core';
import { DataService } from '@vendure/admin-ui/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
import * as i2 from "@vendure/admin-ui/core";
export class AssetResolver extends BaseEntityResolver {
    constructor(router, dataService) {
        super(router, {
            __typename: 'Asset',
            id: '',
            createdAt: '',
            updatedAt: '',
            name: '',
            type: AssetType.IMAGE,
            fileSize: 0,
            mimeType: '',
            width: 0,
            height: 0,
            source: '',
            preview: '',
            focalPoint: null,
        }, id => dataService.product.getAsset(id).mapStream(data => data.asset));
    }
}
AssetResolver.ɵprov = i0.ɵɵdefineInjectable({ factory: function AssetResolver_Factory() { return new AssetResolver(i0.ɵɵinject(i1.Router), i0.ɵɵinject(i2.DataService)); }, token: AssetResolver, providedIn: "root" });
AssetResolver.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
AssetResolver.ctorParameters = () => [
    { type: Router },
    { type: DataService }
];
