// @vendure/admin-ui/esm2015/catalog/providers/routing/facet-resolver.js
import { Injectable } from '@angular/core';
import { Router } from '@angular/router';
import { BaseEntityResolver } from '@vendure/admin-ui/core';
import { getDefaultUiLanguage } from '@vendure/admin-ui/core';
import { DataService } from '@vendure/admin-ui/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
import * as i2 from "@vendure/admin-ui/core";
export class FacetResolver extends BaseEntityResolver {
    constructor(router, dataService) {
        super(router, {
            __typename: 'Facet',
            id: '',
            createdAt: '',
            updatedAt: '',
            isPrivate: false,
            languageCode: getDefaultUiLanguage(),
            name: '',
            code: '',
            translations: [],
            values: [],
        }, (id) => dataService.facet.getFacet(id).mapStream((data) => data.facet));
    }
}
FacetResolver.ɵprov = i0.ɵɵdefineInjectable({ factory: function FacetResolver_Factory() { return new FacetResolver(i0.ɵɵinject(i1.Router), i0.ɵɵinject(i2.DataService)); }, token: FacetResolver, providedIn: "root" });
FacetResolver.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
FacetResolver.ctorParameters = () => [
    { type: Router },
    { type: DataService }
];
