// @vendure/admin-ui/esm2015/catalog/providers/routing/product-variants-resolver.js
import { Injectable } from '@angular/core';
import { Router } from '@angular/router';
import { BaseEntityResolver } from '@vendure/admin-ui/core';
import { DataService } from '@vendure/admin-ui/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
import * as i2 from "@vendure/admin-ui/core";
export class ProductVariantsResolver extends BaseEntityResolver {
    constructor(router, dataService) {
        super(router, {
            __typename: 'Product',
            id: '',
            createdAt: '',
            updatedAt: '',
            name: '',
            optionGroups: [],
            variants: [],
        }, id => dataService.product.getProductVariantsOptions(id).mapStream(data => data.product));
    }
}
ProductVariantsResolver.ɵprov = i0.ɵɵdefineInjectable({ factory: function ProductVariantsResolver_Factory() { return new ProductVariantsResolver(i0.ɵɵinject(i1.Router), i0.ɵɵinject(i2.DataService)); }, token: ProductVariantsResolver, providedIn: "root" });
ProductVariantsResolver.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
ProductVariantsResolver.ctorParameters = () => [
    { type: Router },
    { type: DataService }
];
