// @vendure/admin-ui/esm2015/catalog/components/confirm-variant-deletion-dialog/confirm-variant-deletion-dialog.component.js
import { ChangeDetectionStrategy, Component } from '@angular/core';
export class ConfirmVariantDeletionDialogComponent {
    constructor() {
        this.variants = [];
    }
    confirm() {
        this.resolveWith(true);
    }
    cancel() {
        this.resolveWith();
    }
}
ConfirmVariantDeletionDialogComponent.decorators = [
    { type: Component, args: [{
                selector: 'vdr-confirm-variant-deletion-dialog',
                template: "<ng-template vdrDialogTitle>\r\n    {{ 'catalog.confirm-deletion-of-unused-variants-title' | translate }}\r\n</ng-template>\r\n{{ 'catalog.confirm-deletion-of-unused-variants-body' | translate }}\r\n<ul>\r\n    <li *ngFor=\"let variant of variants\">{{ variant.name }} ({{ variant.sku }})</li>\r\n</ul>\r\n<ng-template vdrDialogButtons>\r\n    <button type=\"button\" class=\"btn\" (click)=\"cancel()\">{{ 'common.cancel' | translate }}</button>\r\n    <button type=\"submit\" (click)=\"confirm()\" class=\"btn btn-primary\">\r\n        {{ 'common.confirm' | translate }}\r\n    </button>\r\n</ng-template>\r\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [""]
            },] }
];
