// @vendure/admin-ui/esm2015/order/components/fulfillment-state-label/fulfillment-state-label.component.js
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
export class FulfillmentStateLabelComponent {
    get chipColorType() {
        switch (this.state) {
            case 'Pending':
            case 'Shipped':
                return 'warning';
            case 'Delivered':
                return 'success';
            case 'Cancelled':
                return 'error';
        }
    }
}
FulfillmentStateLabelComponent.decorators = [
    { type: Component, args: [{
                selector: 'vdr-fulfillment-state-label',
                template: "<vdr-chip [title]=\"'order.payment-state' | translate\" [colorType]=\"chipColorType\">\r\n    <clr-icon shape=\"check-circle\" *ngIf=\"state === 'Delivered'\"></clr-icon>\r\n    {{ state | stateI18nToken | translate }}\r\n</vdr-chip>\r\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [":host{font-size:14px}\n"]
            },] }
];
FulfillmentStateLabelComponent.propDecorators = {
    state: [{ type: Input }]
};
