// @vendure/admin-ui/esm2015/core/shared/components/form-field/form-field-control.directive.js
import { Directive, ElementRef, Optional } from '@angular/core';
import { NgControl } from '@angular/forms';
// tslint:disable:directive-selector
export class FormFieldControlDirective {
    constructor(elementRef, formControlName) {
        this.elementRef = elementRef;
        this.formControlName = formControlName;
    }
    get valid() {
        return !!this.formControlName && !!this.formControlName.valid;
    }
    get touched() {
        return !!this.formControlName && !!this.formControlName.touched;
    }
    setReadOnly(value) {
        const input = this.elementRef.nativeElement;
        if (isSelectElement(input)) {
            input.disabled = value;
        }
        else {
            input.readOnly = value;
        }
    }
}
FormFieldControlDirective.decorators = [
    { type: Directive, args: [{ selector: 'input, textarea, select' },] }
];
FormFieldControlDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: NgControl, decorators: [{ type: Optional }] }
];
function isSelectElement(value) {
    return value.hasOwnProperty('selectedIndex');
}
