// @vendure/admin-ui/esm2015/core/shared/components/facet-value-chip/facet-value-chip.component.js
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output } from '@angular/core';
export class FacetValueChipComponent {
    constructor() {
        this.removable = true;
        this.displayFacetName = true;
        this.remove = new EventEmitter();
    }
}
FacetValueChipComponent.decorators = [
    { type: Component, args: [{
                selector: 'vdr-facet-value-chip',
                template: "<vdr-chip\r\n    [icon]=\"removable ? 'times' : undefined\"\r\n    [colorFrom]=\"facetValue.facet.name\"\r\n    (iconClick)=\"remove.emit()\"\r\n    [title]=\"facetValue.facet.name + ' - ' + facetValue.name\"\r\n>\r\n    <span *ngIf=\"displayFacetName\" class=\"facet-name\">{{ facetValue.facet.name }}</span>\r\n    {{ facetValue.name }}\r\n</vdr-chip>\r\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [":host{display:inline-block}.facet-name{color:var(--color-grey-100);text-transform:uppercase;font-size:10px;margin-right:3px;height:11px}\n"]
            },] }
];
FacetValueChipComponent.propDecorators = {
    facetValue: [{ type: Input }],
    removable: [{ type: Input }],
    displayFacetName: [{ type: Input }],
    remove: [{ type: Output }]
};
