// @vendure/admin-ui/esm2015/core/shared/components/simple-dialog/simple-dialog.component.js
import { ChangeDetectionStrategy, Component } from '@angular/core';
/**
 * Used by ModalService.dialog() to host a generic configurable modal dialog.
 */
export class SimpleDialogComponent {
    constructor() {
        this.title = '';
        this.body = '';
        this.translationVars = {};
        this.buttons = [];
    }
}
SimpleDialogComponent.decorators = [
    { type: Component, args: [{
                selector: 'vdr-simple-dialog',
                template: "<ng-template vdrDialogTitle>{{ title | translate:translationVars }}</ng-template>\r\n{{ body | translate:translationVars }}\r\n<ng-template vdrDialogButtons>\r\n    <ng-container *ngFor=\"let button of buttons\">\r\n        <button\r\n            class=\"btn\"\r\n            [class.btn-primary]=\"button.type === 'primary'\"\r\n            [class.btn-danger]=\"button.type === 'danger'\"\r\n            (click)=\"resolveWith(button.returnValue)\"\r\n        >\r\n            {{ button.label | translate:translationVars }}\r\n        </button>\r\n    </ng-container>\r\n</ng-template>\r\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [""]
            },] }
];
