// @vendure/admin-ui/esm2015/core/shared/dynamic-form-inputs/relation-form-input/relation-card/relation-card.component.js
import { EventEmitter } from '@angular/core';
import { ChangeDetectionStrategy, Component, ContentChild, Directive, Input, Output, TemplateRef, } from '@angular/core';
export class RelationCardPreviewDirective {
}
RelationCardPreviewDirective.decorators = [
    { type: Directive, args: [{
                selector: '[vdrRelationCardPreview]',
            },] }
];
export class RelationCardDetailDirective {
}
RelationCardDetailDirective.decorators = [
    { type: Directive, args: [{
                selector: '[vdrRelationCardDetail]',
            },] }
];
export class RelationCardComponent {
    constructor() {
        this.removable = true;
        this.select = new EventEmitter();
        this.remove = new EventEmitter();
    }
}
RelationCardComponent.decorators = [
    { type: Component, args: [{
                selector: 'vdr-relation-card',
                template: "<div class=\"flex\">\r\n    <ng-container *ngIf=\"entity; else placeholder\">\r\n        <div class=\"preview\">\r\n            <ng-container *ngTemplateOutlet=\"previewTemplate; context: { entity: entity }\"></ng-container>\r\n        </div>\r\n        <div class=\"detail\">\r\n            <div class=\"pl3\">\r\n                <ng-container *ngTemplateOutlet=\"detailTemplate; context: { entity: entity }\"></ng-container>\r\n            </div>\r\n            <button *ngIf=\"!readonly\" class=\"btn btn-sm btn-link\" (click)=\"select.emit()\">\r\n                <clr-icon shape=\"link\"></clr-icon> {{ 'common.change-selection' | translate }}\r\n            </button>\r\n            <button *ngIf=\"!readonly && removable\" class=\"btn btn-sm btn-link\" (click)=\"remove.emit()\">\r\n                <clr-icon shape=\"times\"></clr-icon> {{ 'common.remove' | translate }}\r\n            </button>\r\n        </div>\r\n    </ng-container>\r\n    <ng-template #placeholder>\r\n        <div class=\"preview\">\r\n            <div class=\"placeholder\" (click)=\"!readonly && select.emit()\">\r\n                <clr-icon [attr.shape]=\"placeholderIcon\" size=\"50\"></clr-icon>\r\n            </div>\r\n        </div>\r\n        <div class=\"detail\">\r\n            <div class=\"pl3 not-set\">{{ 'common.not-set' | translate }}</div>\r\n            <button *ngIf=\"!readonly\" class=\"btn btn-sm btn-link\" (click)=\"select.emit()\">\r\n                <clr-icon shape=\"link\"></clr-icon> {{ selectLabel }}\r\n            </button>\r\n        </div>\r\n    </ng-template>\r\n</div>\r\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [":host{display:block;min-width:300px}.placeholder{color:var(--color-grey-300)}.not-set{color:var(--color-grey-300)}.detail{flex:1;overflow:hidden}.name{white-space:nowrap;overflow:hidden;text-overflow:ellipsis}\n"]
            },] }
];
RelationCardComponent.propDecorators = {
    entity: [{ type: Input }],
    placeholderIcon: [{ type: Input }],
    selectLabel: [{ type: Input }],
    readonly: [{ type: Input }],
    removable: [{ type: Input }],
    select: [{ type: Output }],
    remove: [{ type: Output }],
    previewTemplate: [{ type: ContentChild, args: [RelationCardPreviewDirective, { read: TemplateRef },] }],
    detailTemplate: [{ type: ContentChild, args: [RelationCardDetailDirective, { read: TemplateRef },] }]
};
