// @vendure/admin-ui/esm2015/core/shared/components/asset-preview-links/asset-preview-links.component.js
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
export class AssetPreviewLinksComponent {
    constructor() {
        this.sizes = ['tiny', 'thumb', 'small', 'medium', 'large', 'full'];
    }
}
AssetPreviewLinksComponent.decorators = [
    { type: Component, args: [{
                selector: 'vdr-asset-preview-links',
                template: "<vdr-dropdown>\r\n    <button class=\"btn btn-link\" vdrDropdownTrigger>\r\n        <clr-icon shape=\"link\"></clr-icon> {{ 'catalog.asset-preview-links' | translate }}<clr-icon shape=\"caret\" dir=\"down\"></clr-icon>\r\n    </button>\r\n    <vdr-dropdown-menu vdrPosition=\"bottom-left\">\r\n        <a\r\n            *ngFor=\"let size of sizes\"\r\n            [href]=\"asset | assetPreview: size\"\r\n            [title]=\"asset | assetPreview: size\"\r\n            target=\"_blank\"\r\n            class=\"asset-preview-link\"\r\n            vdrDropdownItem\r\n        >\r\n            <vdr-chip><clr-icon shape=\"link\"></clr-icon> {{ 'asset.preview' | translate }}: {{ size }}</vdr-chip>\r\n        </a>\r\n    </vdr-dropdown-menu></vdr-dropdown\r\n>\r\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [".asset-preview-link{font-size:12px}\n"]
            },] }
];
AssetPreviewLinksComponent.propDecorators = {
    asset: [{ type: Input }]
};
