// @vendure/admin-ui/esm2015/core/shared/components/asset-preview-dialog/asset-preview-dialog.component.js
import { ChangeDetectionStrategy, Component } from '@angular/core';
import { of } from 'rxjs';
import { mergeMap } from 'rxjs/operators';
import { DataService } from '../../../data/providers/data.service';
export class AssetPreviewDialogComponent {
    constructor(dataService) {
        this.dataService = dataService;
    }
    ngOnInit() {
        this.assetWithTags$ = of(this.asset).pipe(mergeMap(asset => {
            if (this.hasTags(asset)) {
                return of(asset);
            }
            else {
                // tslint:disable-next-line:no-non-null-assertion
                return this.dataService.product.getAsset(asset.id).mapSingle(data => data.asset);
            }
        }));
    }
    hasTags(asset) {
        return asset.hasOwnProperty('tags');
    }
}
AssetPreviewDialogComponent.decorators = [
    { type: Component, args: [{
                selector: 'vdr-asset-preview-dialog',
                template: "<ng-template vdrDialogTitle>\r\n    <div class=\"title-row\">\r\n        {{ asset.name }}\r\n    </div>\r\n</ng-template>\r\n\r\n<vdr-asset-preview\r\n    *ngIf=\"assetWithTags$ | async as assetWithTags\"\r\n    [asset]=\"assetWithTags\"\r\n    (assetChange)=\"assetChanges = $event\"\r\n    (editClick)=\"resolveWith()\"\r\n></vdr-asset-preview>\r\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [":host{height:70vh}.update-button.hidden{visibility:hidden}\n"]
            },] }
];
AssetPreviewDialogComponent.ctorParameters = () => [
    { type: DataService }
];
