// @vendure/admin-ui/esm2015/core/providers/custom-detail-component/custom-detail-component.service.js
import { APP_INITIALIZER, Injectable } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * @description
 * Registers a {@link CustomDetailComponent} to be placed in a given location. This allows you
 * to embed any type of custom Angular component in the entity detail pages of the Admin UI.
 *
 * @docsCategory custom-detail-components
 */
export function registerCustomDetailComponent(config) {
    return {
        provide: APP_INITIALIZER,
        multi: true,
        useFactory: (customDetailComponentService) => () => {
            customDetailComponentService.registerCustomDetailComponent(config);
        },
        deps: [CustomDetailComponentService],
    };
}
export class CustomDetailComponentService {
    constructor() {
        this.customDetailComponents = new Map();
    }
    registerCustomDetailComponent(config) {
        var _a;
        if (this.customDetailComponents.has(config.locationId)) {
            (_a = this.customDetailComponents.get(config.locationId)) === null || _a === void 0 ? void 0 : _a.push(config);
        }
        else {
            this.customDetailComponents.set(config.locationId, [config]);
        }
    }
    getCustomDetailComponentsFor(locationId) {
        var _a;
        return (_a = this.customDetailComponents.get(locationId)) !== null && _a !== void 0 ? _a : [];
    }
}
CustomDetailComponentService.ɵprov = i0.ɵɵdefineInjectable({ factory: function CustomDetailComponentService_Factory() { return new CustomDetailComponentService(); }, token: CustomDetailComponentService, providedIn: "root" });
CustomDetailComponentService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
