// @vendure/admin-ui/esm2015/order/providers/routing/order-resolver.js
import { Injectable } from '@angular/core';
import { Router } from '@angular/router';
import { BaseEntityResolver } from '@vendure/admin-ui/core';
import { DataService } from '@vendure/admin-ui/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
import * as i2 from "@vendure/admin-ui/core";
/**
 * Resolves the id from the path into a Customer entity.
 */
export class OrderResolver extends BaseEntityResolver {
    constructor(router, dataService) {
        super(router, {
            __typename: 'Order',
            id: '',
            code: '',
            createdAt: '',
            updatedAt: '',
            total: 0,
        }, id => dataService.order.getOrder(id).mapStream(data => data.order));
    }
}
OrderResolver.ɵprov = i0.ɵɵdefineInjectable({ factory: function OrderResolver_Factory() { return new OrderResolver(i0.ɵɵinject(i1.Router), i0.ɵɵinject(i2.DataService)); }, token: OrderResolver, providedIn: "root" });
OrderResolver.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
OrderResolver.ctorParameters = () => [
    { type: Router },
    { type: DataService }
];
