// @vendure/admin-ui/esm2015/dashboard/default-widgets.js
import { marker as _ } from '@biesbjerg/ngx-translate-extract-marker';
import { Permission, } from '@vendure/admin-ui/core';
import { LatestOrdersWidgetComponent } from './widgets/latest-orders-widget/latest-orders-widget.component';
import { OrderSummaryWidgetComponent } from './widgets/order-summary-widget/order-summary-widget.component';
import { TestWidgetComponent } from './widgets/test-widget/test-widget.component';
import { WelcomeWidgetComponent } from './widgets/welcome-widget/welcome-widget.component';
export const DEFAULT_DASHBOARD_WIDGET_LAYOUT = [
    { id: 'welcome', width: 12 },
    { id: 'orderSummary', width: 6 },
    { id: 'latestOrders', width: 6 },
];
const ɵ0 = () => WelcomeWidgetComponent, ɵ1 = () => OrderSummaryWidgetComponent, ɵ2 = () => LatestOrdersWidgetComponent, ɵ3 = () => TestWidgetComponent;
export const DEFAULT_WIDGETS = {
    welcome: {
        loadComponent: ɵ0,
    },
    orderSummary: {
        title: _('dashboard.orders-summary'),
        loadComponent: ɵ1,
        requiresPermissions: [Permission.ReadOrder],
    },
    latestOrders: {
        title: _('dashboard.latest-orders'),
        loadComponent: ɵ2,
        supportedWidths: [6, 8, 12],
        requiresPermissions: [Permission.ReadOrder],
    },
    testWidget: {
        title: 'Test Widget',
        loadComponent: ɵ3,
    },
};
export { ɵ0, ɵ1, ɵ2, ɵ3 };
