// @vendure/admin-ui/esm2015/core/shared/components/table-row-action/table-row-action.component.js
import { Component, Input } from '@angular/core';
/**
 * A button link to be used as actions in rows of a table.
 */
export class TableRowActionComponent {
    constructor() {
        this.disabled = false;
    }
}
TableRowActionComponent.decorators = [
    { type: Component, args: [{
                selector: 'vdr-table-row-action',
                template: "<ng-container *ngIf=\"!disabled; else: disabledLink\">\r\n    <a class=\"btn btn-link btn-sm\" [routerLink]=\"linkTo\">\r\n        <clr-icon [attr.shape]=\"iconShape\"></clr-icon>\r\n        {{ label }}\r\n    </a>\r\n</ng-container>\r\n<ng-template #disabledLink>\r\n    <button class=\"btn btn-link btn-sm\" disabled>\r\n        <clr-icon [attr.shape]=\"iconShape\"></clr-icon>\r\n        {{ label }}\r\n    </button>\r\n</ng-template>\r\n",
                styles: ["a.btn{margin:0}\n"]
            },] }
];
TableRowActionComponent.propDecorators = {
    linkTo: [{ type: Input }],
    label: [{ type: Input }],
    iconShape: [{ type: Input }],
    disabled: [{ type: Input }]
};
