// @vendure/admin-ui/esm2015/core/shared/providers/routing/can-deactivate-detail-guard.js
import { Injectable } from '@angular/core';
import { Router } from '@angular/router';
import { marker as _ } from '@biesbjerg/ngx-translate-extract-marker';
import { map } from 'rxjs/operators';
import { ModalService } from '../../../providers/modal/modal.service';
export class CanDeactivateDetailGuard {
    constructor(modalService, router) {
        this.modalService = modalService;
        this.router = router;
    }
    canDeactivate(component, currentRoute, currentState, nextState) {
        if (!component.canDeactivate()) {
            return this.modalService
                .dialog({
                title: _('common.confirm-navigation'),
                body: _('common.there-are-unsaved-changes'),
                buttons: [
                    { type: 'danger', label: _('common.discard-changes'), returnValue: true },
                    { type: 'primary', label: _('common.cancel-navigation'), returnValue: false },
                ],
            })
                .pipe(map(result => !!result));
        }
        else {
            return true;
        }
    }
}
CanDeactivateDetailGuard.decorators = [
    { type: Injectable }
];
CanDeactivateDetailGuard.ctorParameters = () => [
    { type: ModalService },
    { type: Router }
];
