// @vendure/admin-ui/esm2015/core/shared/components/title-input/title-input.component.js
import { ChangeDetectionStrategy, Component, HostBinding, Input } from '@angular/core';
export class TitleInputComponent {
    constructor() {
        this.readonly = false;
    }
}
TitleInputComponent.decorators = [
    { type: Component, args: [{
                selector: 'vdr-title-input',
                template: "<ng-content></ng-content>\r\n<div class=\"edit-icon\" *ngIf=\"!readonly\">\r\n    <clr-icon shape=\"edit\"></clr-icon>\r\n</div>\r\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [":host{display:block;position:relative}:host ::ng-deep input{font-size:18px;color:var(--clr-p1-color);max-width:100%}:host ::ng-deep input:not(:focus){border-color:transparent!important;background-color:var(--color-component-bg-100)!important}:host ::ng-deep input:focus{background-color:var(--clr-global-app-background)}:host ::ng-deep .clr-control-container{max-width:100%}:host ::ng-deep .clr-input-wrapper{max-width:100%!important}:host .edit-icon{position:absolute;right:8px;top:6px;color:var(--color-grey-400);opacity:0;transition:opacity .2s}:host:hover ::ng-deep input:not(:focus){background-color:var(--color-component-bg-200)!important}:host:hover .edit-icon{opacity:1}:host.readonly .edit-icon{display:none}:host.readonly:hover ::ng-deep input:not(:focus){background-color:var(--color-grey-200)!important}\n"]
            },] }
];
TitleInputComponent.propDecorators = {
    readonly: [{ type: HostBinding, args: ['class.readonly',] }, { type: Input }]
};
