// @vendure/admin-ui/esm2015/marketing/providers/routing/promotion-resolver.js
import { Injectable } from '@angular/core';
import { Router } from '@angular/router';
import { BaseEntityResolver, DataService } from '@vendure/admin-ui/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
import * as i2 from "@vendure/admin-ui/core";
/**
 * Resolves the id from the path into a Customer entity.
 */
export class PromotionResolver extends BaseEntityResolver {
    constructor(router, dataService) {
        super(router, {
            __typename: 'Promotion',
            id: '',
            createdAt: '',
            updatedAt: '',
            name: '',
            enabled: false,
            conditions: [],
            actions: [],
        }, id => dataService.promotion.getPromotion(id).mapStream(data => data.promotion));
    }
}
PromotionResolver.ɵprov = i0.ɵɵdefineInjectable({ factory: function PromotionResolver_Factory() { return new PromotionResolver(i0.ɵɵinject(i1.Router), i0.ɵɵinject(i2.DataService)); }, token: PromotionResolver, providedIn: "root" });
PromotionResolver.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
PromotionResolver.ctorParameters = () => [
    { type: Router },
    { type: DataService }
];
