// @vendure/admin-ui/esm2015/core/shared/components/entity-info/entity-info.component.js
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
export class EntityInfoComponent {
    constructor() {
        this.small = false;
    }
}
EntityInfoComponent.decorators = [
    { type: Component, args: [{
                selector: 'vdr-entity-info',
                template: "<vdr-dropdown *ngIf=\"entity.id\">\r\n    <button class=\"btn btn-icon btn-link info-button\" [class.btn-sm]=\"small\" vdrDropdownTrigger>\r\n        <clr-icon shape=\"info-standard\"></clr-icon>\r\n    </button>\r\n    <vdr-dropdown-menu>\r\n        <div class=\"entity-info\">\r\n            <vdr-labeled-data [label]=\"'common.ID' | translate\">\r\n                {{ entity.id }}\r\n            </vdr-labeled-data>\r\n            <vdr-labeled-data *ngIf=\"entity.createdAt\" [label]=\"'common.created-at' | translate\">\r\n                {{ entity.createdAt | localeDate: 'medium' }}\r\n            </vdr-labeled-data>\r\n            <vdr-labeled-data *ngIf=\"entity.updatedAt\" [label]=\"'common.updated-at' | translate\">\r\n                {{ entity.updatedAt | localeDate: 'medium' }}\r\n            </vdr-labeled-data>\r\n        </div>\r\n    </vdr-dropdown-menu>\r\n</vdr-dropdown>\r\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [".info-button{color:var(--color-icon-button)}.entity-info{margin:0 12px}\n"]
            },] }
];
EntityInfoComponent.propDecorators = {
    small: [{ type: Input }],
    entity: [{ type: Input }]
};
