// @vendure/admin-ui/esm2015/core/shared/components/custom-detail-component-host/custom-detail-component-host.component.js
import { ChangeDetectionStrategy, Component, ComponentFactoryResolver, Input, ViewContainerRef, } from '@angular/core';
import { CustomDetailComponentService } from '../../../providers/custom-detail-component/custom-detail-component.service';
export class CustomDetailComponentHostComponent {
    constructor(viewContainerRef, componentFactoryResolver, customDetailComponentService) {
        this.viewContainerRef = viewContainerRef;
        this.componentFactoryResolver = componentFactoryResolver;
        this.customDetailComponentService = customDetailComponentService;
        this.componentRefs = [];
    }
    ngOnInit() {
        const customComponents = this.customDetailComponentService.getCustomDetailComponentsFor(this.locationId);
        for (const config of customComponents) {
            const factory = this.componentFactoryResolver.resolveComponentFactory(config.component);
            const componentRef = this.viewContainerRef.createComponent(factory);
            componentRef.instance.entity$ = this.entity$;
            componentRef.instance.detailForm = this.detailForm;
            this.componentRefs.push(componentRef);
        }
    }
    ngOnDestroy() {
        for (const ref of this.componentRefs) {
            ref.destroy();
        }
    }
}
CustomDetailComponentHostComponent.decorators = [
    { type: Component, args: [{
                selector: 'vdr-custom-detail-component-host',
                template: "<vdr-ui-extension-point [locationId]=\"locationId\" api=\"detailComponent\"></vdr-ui-extension-point>\r\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [""]
            },] }
];
CustomDetailComponentHostComponent.ctorParameters = () => [
    { type: ViewContainerRef },
    { type: ComponentFactoryResolver },
    { type: CustomDetailComponentService }
];
CustomDetailComponentHostComponent.propDecorators = {
    locationId: [{ type: Input }],
    entity$: [{ type: Input }],
    detailForm: [{ type: Input }]
};
