// @vendure/admin-ui/esm2015/order/components/payment-state-label/payment-state-label.component.js
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
export class PaymentStateLabelComponent {
    get chipColorType() {
        switch (this.state) {
            case 'Authorized':
                return 'warning';
            case 'Settled':
                return 'success';
            case 'Declined':
            case 'Cancelled':
                return 'error';
        }
    }
}
PaymentStateLabelComponent.decorators = [
    { type: Component, args: [{
                selector: 'vdr-payment-state-label',
                template: "<vdr-chip [title]=\"'order.payment-state' | translate\" [colorType]=\"chipColorType\">\r\n    <clr-icon shape=\"check-circle\" *ngIf=\"state === 'Settled'\"></clr-icon>\r\n    {{ state | stateI18nToken | translate }}\r\n</vdr-chip>\r\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [":host{font-size:14px}\n"]
            },] }
];
PaymentStateLabelComponent.propDecorators = {
    state: [{ type: Input }]
};
